code/iaas/config-item/common/src/main/java/io/cattle/platform/configitem/version/impl/ConfigUpdatePublisher.java

```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    private void processRequest(WorkItem item) {
+        CollectionUtils.addToMap(this.waiters, item.key, item, ArrayList.class);
+    }
+
+    protected EventCallOptions defaultOptions() {
+        EventCallOptions options = new EventCallOptions(RETRY.get(), TIMEOUT.get()).withProgress(new EventProgress() {
+            @Override
+            public void progress(Event event) {
+                ConfigItemStatusManagerImpl.logResponse(null, event);
+            }
+        });
+
+        return options;
+    }
+
+    protected ListenableFuture<? extends Event> call(Client client, ConfigUpdate event) {
+        EventCallOptions options = defaultOptions();
+        if (client.getResourceType() == Agent.class) {
+            RemoteAgent agent = agentLocator.lookupAgent(client.getResourceId());
+            return agent.call(event, options);
+        }
+
+        return eventService.call(event, options);
+    }
+
+
+    protected Map<String, ItemVersion> getApplied(Client client) {
+        return configItemStatusDao.getApplied(client);
+    }
+
+    protected boolean satisfies(boolean isRequest, ConfigUpdate request, Map<String, ItemVersion> applied) {
+        if (request.getData() == null) {
+            return true;
+        }
+
+        for (ConfigUpdateItem item : request.getData().getItems()) {
+            ItemVersion version = applied.get(item.getName());
+            if (version == null) {
+                if (log.isTraceEnabled() && !isRequest) {
+                    log.info("\t\tUnsatified item {} [{}] is not assigned", item.hashCode(), item.getName());
+                }
+                return false;
+            }
+
+            if (item.getRequestedVersion() == null) {
+                if (isRequest) {
+                    return false;
+                } else {
+                    continue;
+                }
+            }
+
+            if (version.getRevision() < item.getRequestedVersion()) {
+                if (log.isTraceEnabled() && !isRequest) {
+                    log.info("\t\tUnsatified item {} [{}] [{}<{}]", item.hashCode(), item.getName(), version.getRevision(), item.getRequestedVersion());
+                }
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected void reply(WorkItem item, Event response, Throwable t) {
+        if (t != null) {
+            item.future.setException(t);
+            return;
+        }
+
+        EventVO<Object> event = EventVO.reply(item.request);
+        if (response != null) {
+            EventUtils.copyTransitioning(response, event);
+            event.setData(response.getData());
+        }
+
+        log.info("\t\tFinished work item {} [{}]", item.key, item.hashCode());
+        item.future.set(event);
+    }
+
+    @Override
+    public void start() {
+        running = true;
+        executorService.scheduleWithFixedDelay(this, 0, 2, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void stop() {
```

Comment:
When/how does stop() get called by the framework?

I see doRun has a loop that will basically go until stop is called and the executor service will execute (start?) this every two seconds. So, how does run A stop so that run B 2 seconds later can execute?

I'm sure this is just something I don't know about the framework.